### ISSUE TITLE
Update permissions for Android API > 10

### ISSUE THREAD

**Issue by ST-Music** [gogins/csound-android#18]
Just installed new version. Was necessary to change directory to "/storage/emulated/0/Music/Csound6AndroidExamples". Can now render wav files but cannot load them even when using full path. Cannot load any example files. Help file not available.

Reinstalled previous version & wav files load ok.

**Comment by gogins** [comment]
Are you talking about version code 32? (shown in the Messages tab on
startup).

-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by ST-Music** [comment]
Hi Michael, had tested it at about 1 pm. Will try again later. Have you
tested it with reading in a wav file? If so could you possibly post a small
ex. that works for you?

In the future I'll remember to always include version codes.

**Comment by gogins** [comment]
In case version 32 is not yet available in the Play Store, I have created a
beta release that you can download from here:
https://github.com/gogins/csound-android/releases/tag/untagged-d010345952eb9316afec
.

I will test the file reading and get back to you about that.

-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by ST-Music** [comment]
Thanks, that's great. Play shows last update on Aug. 25. The link you sent
gives a 404 so I'll just sit tight for now. And thanks again for taking the
time to do this, I'm sure it has it's frustrations.

**Comment by gogins** [comment]
I copied the attached file to the Music directory on my tablet, and it ran
just fine and saved three soundfiles, as it was supposed to.

I used the link in my email to get to my release on GitHub just fine. From
it I downloaded the .apk, and the size looks right.

Possibly you specified a file in a subdirectory that does not exist?

Regards,
Mike

-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by ST-Music** [comment]
This is weird. First, there is no attached file. Secondly, the link still
did not work. I even copy & pasted it into a browser and still get a 404. As
for the directory issue, the path was fine. As soon as I reinstalled the
old app the same csd loaded the wav no problem. I triple checked the file location & confirmed the file was in that directory. I have a csd example if you
would like to check it out. If you change the filename to fox.wav it should
be fine although you may have to adjust the path:

https://www.dropbox.com/scl/fi/h01v6d8bg3j23136uyaz4/split_wav_ex_2.csd?rlkey=88f248ike5gs408qt5pneyhyv&dl=0

Worth noting,  I can render csd files (that don't use a wav file) to wav files fine using the same directory. 

(-o but without parenthesis) /storage/emulated/0/Music/Csound6AndroidExamples/render.wav

When I do, using a file explorer, I find the rendered file no problem, alongside the file that fails to load in other csd. And as mentioned, the previous version of Csound I reverted back to (temporarily) plays the same csd without issues, so it can't be a faulty path.

Using this link:
https://github.com/gogins/csound-android/releases/tag/untagged-d010345952eb9316afec

I get this error. As you can see I am logged into github so that shouldn't
be the issue:

https://www.dropbox.com/scl/fi/l8wx73pscgk75204x7o6q/Screenshot_20230826_171405_Chrome.jpg?rlkey=4uttqxjv2xzm4ctvxbvw2738r&dl=0


And FWIW the upgrade is no longer available via Play, only the previous version from 2021.

**Comment by gogins** [comment]
I guess your email blocked the attachment. Here it is in the email, below.
You can try this URL for the download:
https://github.com/gogins/csound-android/releases

<CsoundSynthesizer>
<CsOptions>
; Select audio/midi flags here according to platform
-odac     ;;;realtime audio out
;-iadc    ;;;uncomment -iadc if realtime audio input is needed too
; For Non-realtime ouput leave only the line below:
; -o fout.wav -W ;;; for file output any platform
</CsOptions>
<CsInstruments>

; Required settings for WebAudio:

sr = 48000
ksmps = 128
nchnls = 2
nchnls_i = 1

; sr = 48000
; ksmps = 128
; nchnls = 2
0dbfs  = 1

giSine ftgen 0, 0, 2^10, 10, 1

instr 1

asig poscil3 .5, 880, giSine
;write a raw file: 32 bits with header
     fout "/storage/emulated/0/Music/fout_880.wav", 15, asig
     outs asig, asig

endin

instr 2

klfo lfo 1, 2, 0
asig poscil3 .5*klfo, 220, giSine
;write an aiff file: 32 bits with header
     fout "/storage/emulated/0/Music/fout_aif.aiff", 25, asig
;        fout "fout_all3.wav", 14, asig
     outs asig, asig

endin

instr 99 ;read the stereo csound output buffer

allL, allR monitor
;write the output of csound to an audio file
;to a wav file: 16 bits with header
           fout "/storage/emulated/0/Music/fout_all.wav", 14, allL, allR

endin
</CsInstruments>
<CsScore>

i 1 0 2
i 2 0 3
i 99 0 3
e
</CsScore>
</CsoundSynthesizer>


-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by ST-Music** [comment]
Unfortunately this just isn't my day it seems =)

When I click the email link, or copy and paste it into a browser, I only
see the old version of Csound. You can see by the date:
https://www.dropbox.com/scl/fi/7k9p7uueoor4riup9icdd/Screenshot_20230826_195146_Samsung-Internet.jpg?rlkey=9j6l1b35ydy0ifx29hpq4o9k6&dl=0

When I scroll down you can see that the update is not available:
https://www.dropbox.com/scl/fi/bbelhh00h1sg8qfqna6hz/Screenshot_20230826_195704_Chrome.jpg?rlkey=6t4nksergz15bq05jr3ufhmtj&dl=0

Perhaps github hasn't yet processed the link yet (I've had it take an hour
before), I don't know.

So unless you can upload it somewhere like Google Drive or Dropbox I can't
test it as I didn't save the updated apk assuming it would still be
available on Play. Normally I backup all apk files as problematic updates
for other apps is not new to me, I should have done so with the update &
will in the future.

**Comment by gogins** [comment]
Thank you. I thought I had published this release, but it was still a
draft. Could you please try again?

Thanks,
Mike

-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by ST-Music** [comment]
Thanks, that works, downloading it now.

And thank you for the example but unfortunately it only SAVES a file via
fout.

I had no issues with the updated version when saving audio files. I must
reiterate, the issue is that it would not LOAD them.

Have you been able to load wav files from disk? If you do you have an
example.

Meanwhile I'll try testing it again.

**Comment by ST-Music** [comment]
Ok, tried again. This time it loads the file but the playback is flawed and
sounds very odd, barely audible and buzzy. It also reports in the console
that it fails to load the file when in fact it does seem to. I'll try a
with a simplified csd.

Here is the console output:
https://www.dropbox.com/scl/fi/blyt0gcyuwb2l41twquis/Screenshot_20230826_202358.jpg?rlkey=pcmtw67vaccacuiwue8xs0juu&dl=0

**Comment by ST-Music** [comment]
Here is a simple example. It gives an error message, soundin cannot open
etc.

I uninstalled, reinstalled the previous version, and it plays fine.

<CsoundSynthesizer>
<CsOptions>
 -odac
</CsOptions>
;==================================
<CsInstruments>

sr = 48000
ksmps = 32
nchnls = 1
0dbfs  = 1

giTab   ftgen 1, 0, 0, 1, "/storage/emulated/0/Music/fox.wav", 0, 0, 1

instr 1
  out poscil(1, .3, 1)
endin

</CsInstruments>
;==================================
<CsScore>
i1 0 4
</CsScore>
</CsoundSynthesizer>

**Comment by gogins** [comment]
Yes, I have been able to load files. My example follows.

<CsoundSynthesizer>
<CsOptions>
; Select audio/midi flags here according to platform
-odac     ;;;RT audio out
;-iadc    ;;;uncomment -iadc if RT audio input is needed too
; For Non-realtime ouput leave only the line below:
; -o diskin2.wav -W ;;; for file output any platform
</CsOptions>
<CsInstruments>

sr     = 48000
ksmps  = 128
nchnls = 2
0dbfs  = 1

instr 1

ktrans linseg 1, 5, 2, 10, -2
a1     diskin2 "/storage/emulated/0/Music/samples/beats.wav", ktrans, 0, 1,
0, 32
       outs a1, a1

endin

</CsInstruments>
<CsScore>

i 1 0 15
e

</CsScore>
</CsoundSynthesizer>

-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by gogins** [comment]
Csound for Android copies all assets out of the .apk file into the external
public storage directory for music. There are subdirectories in that
directory, and all samples such as fox.wav go into the samples
subdirectory. Try changing "fox.wav" to "samples/fox.wav".

-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by ST-Music** [comment]
Unfortunately it didn't work here.

As you can see from the console output, it fails to load:
https://www.dropbox.com/scl/fi/l2h966o1gwfvcgsi44p1g/Screenshot_20230826_210651.jpg?rlkey=a3tcmn6e3ozvmt09mesvlbpq1&dl=0

This is a Screenshot from Music/samples (you can see that at the top) & you
can see the sample is in the correct directory:
https://www.dropbox.com/scl/fi/1qfwqzl5rx3bmv8cdjlo9/Screenshot_20230826_210743_ASTRO-File-Manager.jpg?rlkey=w3gpxjrp9np8mxd95ie3apyui&dl=0

This occurs on both my Samsung S6 & S20.

You mentioned in the first reply ver code 32. The app is reporting ver code
31, if that matters. However, it reports the Csound library as 6160. Unfortunately as the app stands, installing the new update does not appear to be replacing it.

Perhaps it may be necessary to fully uninstall the prior version before updating it, but I have reservations. If it doesn't work I'm potentially left without the ability to use Csound.

And FWIW, I used your example to test. Once again, reinstalling the original Csound it does work.

**Comment by gogins** [comment]
I'll have to look at this tomorrow or Monday, it's getting late here.

I have a tablet and a phone. Everything works fine on the tablet, which has
Android 10, but I have problems unpacking the assets or loading other files
from the app on the phone, which has a later version of the operating
system. I probably need to do some work on file access and permissions for
Android 11 through 13.

-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by ST-Music** [comment]
No worries, the older version still works fine so I'm good to go. And so
you know I am using Android 13 on my new phone. Unfortunately many devs
have found the new permissions frustrating. Sleep well!

**Comment by gogins** [comment]
The permissions mess is detailed here: https://developer.android.com/training/data-storage/manage-all-files and here: https://support.google.com/googleplay/android-developer/answer/10467955?sjid=8237186783112561804-NA&visit_id=21693163340047-275003872552940314&rd=1.

Reading between the lines, Google decided their own Storage Access Framework (SAF) wasn't working out for them, and replaced it with _much_ more restrictive file access permissions, along with a new MANAGE_EXTERNAL_STORAGE permission. However, if that permission is used, then code written using the SAF doesn't get any new access, so you have to use the File class or native code instead. Also, you have to petition Google to grant your app this permission in the Play Store; apparently this happens you submit a release.

It's possible I'll have to remove the app from the Play Store. I'll prepare a new testing release with MANAGE_EXTERNAL_STORAGE permission and submit it for review; if there's a problem that can't be resolved with a reasonable amount of work, I'll pull the app from the Play Store and publish it only as a release on GitHub.

The problem is that I need MANAGE_EXTERNAL_STORAGE to _programmatically_ copy assets to shared storage. Once they're there, I can just use the SAF (I think).

**Comment by gogins** [comment]
This is a summary of use cases for the Csound for Android app. The summary is not complete, this is just a start, as the Csound library native code, and the Android WebBrowser running JavaScript, do many, many things.

If I have verified that the use case is supported directly by Android, by the Storage Access Framework, by the MediaStore, or by MANAGE_EXTERNAL_STORAGE, I have added A and/or SAF and/or MS and/or MES. I don't know all the answers. MES may correspond to the Files permission in the Android settings, or it may be a combination of reading external storage and writing external storage.

1. When the app starts up, it writes examples, code, and samples from the .apk file to shared storage (MES).
2. The user downloads a .csd file, and a .wav file used by the .csd file as a sample, to shared storage (A).
3. The user opens this piece in the app and inspects it (SAF, MES).
4. The user successfully renders this piece to audio (A).
5. The user edits this piece to write a soundfile to shared storage instead of rendering to audio, and saves the piece to shared storage with a new filename (MES).
6. The user edits this piece and embeds into it an HTML page that the app displays using its builtin WebView. The Web page loads JavaScript code and images from shared storage (MES).
7. The user runs the piece and the app succesfully renders the soundfile to shared storage, while displaying the embedded Web page, which runs JavaScript code on the page to generate notes for Csound to render (MES).
8. The user uploads the new soundfile to the cloud (A).

**Comment by gogins** [comment]
This is funny (https://stackoverflow.com/questions/59000390/android-accessing-files-in-native-c-c-code-with-google-scoped-storage-api) , I'm crying. 

I'm partway there. Without all files access, I can go to the Android settings and enable Files access and other permissions, and that seems to work. But I don't see how to get, e.g., embedded Web pages to load and run code.

**Comment by ST-Music** [comment]
﻿I wish I got the joke but those related to programming often go over my inexperienced head.

A not so quick note - I did post another issue possibly relating to directory access (the fact the "directory" and ftsamplebank opcodes don't seem to be working in the beta2 seems rather coincidental & an odd error message - syntax related, not "files not found").

How you proceed is your decision. These opcodes and those in the 6.16 lib worked fine before. Personally, I don't really care if it goes on Google Play. If you feel feel like just using the previous "app" or build but making an apk that instead includes the 6.18 library, if that's even possible, & distributing only on github, I'd be fine with that. And perhaps easier for you; maybe then issues like getting embedded html pages to work wouldn't be a problem?

Just a thought. All of this is generally over my head.

The previous version worked fine on Android 13 with the exception of crashes which seem to occur at the same rate in beta2. On a purely selfish level I'd just like to be be able to use the 6.18 library. But if it's a choice between using the directory and ftsamplebank opcodes or not, they play a major role in my compositions so I'd end up still generally using the old app, or installing the new on occasion & then reinstalling the old for general use.

Let me know if you read this, I'll delete it so it's not clogging up the thread.
